From 1c34842323a550f3f7f09558e4ab7acc37621d09 Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@st.com>
Date: Thu, 20 Oct 2022 15:59:02 +0200
Subject: [PATCH 0519/1695] plat-stm32mp1: BKPRAM driver support

STM32MP1 SoC embeds a RAM named BKPRAM that is retained during some
low power states or when the platform embeds an internal battery.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git fe30a31b8eae2d68c7a115c0737116cb3e64ac1b plat-stm32mp1: BKPRAM driver support]
Change-Id: I837b985bc28c78dee7b24873780ae52463a7a2d9
Signed-off-by: Etienne Carriere <etienne.carriere@linaro.org>
Signed-off-by: Pascal Paillet <p.paillet@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/358362
Tested-by: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
---
 core/arch/arm/plat-stm32mp1/main.c            | 7 +++++++
 core/arch/arm/plat-stm32mp1/platform_config.h | 7 +++++++
 core/arch/arm/plat-stm32mp1/stm32_util.h      | 1 +
 3 files changed, 15 insertions(+)

diff --git a/core/arch/arm/plat-stm32mp1/main.c b/core/arch/arm/plat-stm32mp1/main.c
index 098de60b6..001e84da8 100644
--- a/core/arch/arm/plat-stm32mp1/main.c
+++ b/core/arch/arm/plat-stm32mp1/main.c
@@ -443,6 +443,13 @@ vaddr_t stm32mp_bkpreg(unsigned int idx)
 	return bkpreg_base() + (idx * sizeof(uint32_t));
 }
 
+vaddr_t stm32mp_bkpsram_base(void)
+{
+	struct io_pa_va base = { .pa = BKPSRAM_BASE };
+
+	return io_pa_or_va(&base, BKPSRAM_SIZE);
+}
+
 static bool __maybe_unused bank_is_valid(unsigned int bank)
 {
 	if (IS_ENABLED(CFG_STM32MP15))
diff --git a/core/arch/arm/plat-stm32mp1/platform_config.h b/core/arch/arm/plat-stm32mp1/platform_config.h
index 12152d9cd..4672edc72 100644
--- a/core/arch/arm/plat-stm32mp1/platform_config.h
+++ b/core/arch/arm/plat-stm32mp1/platform_config.h
@@ -264,6 +264,13 @@
 #define DBGMCU_IDC_REV_ID_MASK		GENMASK_32(31, 16)
 #define DBGMCU_IDC_REV_ID_SHIFT		U(16)
 
+/* BKPSRAM layout */
+#ifdef CFG_STM32MP13
+#define BKPSRAM_SIZE			U(0x2000)
+#else
+#define BKPSRAM_SIZE			U(0x1000)
+#endif
+
 /* SYSRAM layout */
 #ifdef CFG_STM32MP13
 #define SYSRAM_SIZE			0x20000
diff --git a/core/arch/arm/plat-stm32mp1/stm32_util.h b/core/arch/arm/plat-stm32mp1/stm32_util.h
index 996ce12ac..1d58d39fb 100644
--- a/core/arch/arm/plat-stm32mp1/stm32_util.h
+++ b/core/arch/arm/plat-stm32mp1/stm32_util.h
@@ -28,6 +28,7 @@ bool stm32mp_supports_second_core(void);
 
 /* Backup registers and RAM utils */
 vaddr_t stm32mp_bkpreg(unsigned int idx);
+vaddr_t stm32mp_bkpsram_base(void);
 
 /* Get device ID from SYSCFG registers */
 uint32_t stm32mp_syscfg_get_chip_dev_id(void);
-- 
2.47.2

